src/mlpack/core/util/backtrace.cpp

```diff
@@ -0,0 +1,191 @@
+asection *text = 0;	// Strings datastructure.
+#endif
+
+#ifdef HAS_BFD_DL
+Backtrace::Backtrace(int maxDepth)
+{
+  frame.address = NULL;
+  frame.function = "0";
+  frame.file = "0";
+  frame.line = 0;
+  
+  stack.clear();
+  
+  GetAddress(maxDepth);
+}
+#else
+Backtrace::Backtrace()
+{
+  // Dummy constructor
+}
+#endif
+
+#ifdef HAS_BFD_DL
+void Backtrace::GetAddress(int maxDepth)
+{
+  void* trace[maxDepth];
+  int stackDepth = backtrace(trace, maxDepth);
+
+  // Skip first stack frame (points to Backtrace::Backtrace).
+  for (int i = 1; i < stackDepth; i++) 
+  {
+    Dl_info addressHandler;
+    
+    //No backtrace will be printed if no compile flags: -g -rdynamic
+    if(TRACE_CONDITION_1)
+    {
+      return ;
+    }
+    
+    frame.address = addressHandler.dli_saddr;
+    
+    DecodeAddress((long)frame.address);
+  }
+}
+
+void Backtrace::DecodeAddress(long addr)
+{
+  // Check to see if there is anything to descript. If it doesn't, we'll
+  // dump running program.
+  if (!abfd)
+  {
+    char ename[1024];
+    int l = readlink("/proc/self/exe",ename,sizeof(ename));
+    if (l == -1)
+    {
+      perror("Failed to open executable!\n");
+      return;
+    }
+    ename[l] = 0;
+    
+    bfd_init();
+    
+    abfd = bfd_openr(ename, 0);
+    if (!abfd)
+    {
+      perror("bfd_openr failed: ");
+      return;
+    }
+ 
+    bfd_check_format(abfd,bfd_object);
+ 
+    unsigned storage_needed = bfd_get_symtab_upper_bound(abfd);
+    syms = (asymbol **) malloc(storage_needed);
+  
+    text = bfd_get_section_by_name(abfd, ".text");
+   }
+   
+  long offset = addr - text->vma;
+  
+  if (offset > 0)
+  {	
+    if(FIND_LINE)
+    {
+      DemangleFunction();
+      // Save retrieved informations.
+      stack.push_back(frame);
+    }
+  }
+}
+
+void Backtrace::DemangleFunction()
+{
+  int status;
+  char* tmp = abi::__cxa_demangle(frame.function, 0, 0, &status);
+  
+  // If demangling is successful, reallocate 'frame.function' pointer to
+  // demangled name. Else if 'status != 0', leave 'frame.function as it is.
+  if (status == 0)
+  {
+    frame.function = tmp;
+  }
+}
+#else
+void Backtrace::GetAddress(int /* maxDepth */) { }
+void Backtrace::DecodeAddress(long /* address */) { }
+void Backtrace::DemangleFunction() { }
+#endif
+
+std::string Backtrace::ToString()
+{  
+  std::string stackStr;
+  
+#ifdef HAS_BFD_DL  
+  std::ostringstream lineOss;
+  std::ostringstream it;
+
+  if(stack.size() <= 0)
+  {
+    stackStr = "Cannot give backtrace because program was compiled";
+    stackStr += " without: -g -rdynamic\nFor a backtrace,";
+    stackStr += " recompile with: -g -rdynamic.\n";
+    
+    return stackStr;
+  }
+  
+  for(unsigned int i = 0; i < stack.size(); i++)
+  {
+    frame = stack[i];
+    
+    lineOss << frame.line;
+    it << i + 1;
+    
+      stackStr += "[bt]: (" + it.str() + ") "
+	       + frame.file + ":"
+	       + frame.function + ":"
+	       + lineOss.str() + "\n";
```

Comment:
I think it might be useful to print `lineOss.str()` before `frame.function`; this is a bit closer to the output given by utilities like gdb and similar.  So I guess instead of output like

```
[bt]: a.cpp:A():10
```

I think that this would be better:

```
[bt]: a.cpp:10: A()
```
(I think a space between the line number and the function would be helpful for readability too.)